hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/mapreduce/MapReduceHFileSplitterJob.java

```diff
@@ -86,6 +101,152 @@ public void setup(Context context) throws IOException {
+        LOG.error("Error creating rack-aware splits, falling back to standard splits", e);
+        return super.getSplits(context);
+      }
+    }
+    
+    private List<InputSplit> createRackAwareSplits(JobContext context) 
+        throws IOException {
+      Configuration conf = context.getConfiguration();
+      
+      // 1. Get all HFiles from input paths
+      List<FileStatus> files = listStatus(context);
+      Collection<Pair<String, Long>> hfiles = new ArrayList<>();
+      for (FileStatus file : files) {
+        hfiles.add(new Pair<>(file.getPath().toString(), file.getLen()));
+      }
+      
+      // 2. Load pluggable CustomFileGrouper
+      Class<? extends HFileCustomFileGrouper> grouperClass = conf.getClass(
+        CONF_INPUT_FILE_GROUPER_CLASS, 
+        NoopHFileCustomFileGrouper.class, 
+        HFileCustomFileGrouper.class);
+      HFileCustomFileGrouper customFileGrouper = 
+        ReflectionUtils.newInstance(grouperClass, conf);
+      
+      // 3. Load pluggable FileLocationResolver
+      Class<? extends HFileLocationResolver> resolverClass = conf.getClass(
+        CONF_INPUT_FILE_LOCATION_RESOLVER_CLASS,
+        NoopHFileLocationResolver.class,
+        HFileLocationResolver.class);
+      HFileLocationResolver fileLocationResolver = 
+        ReflectionUtils.newInstance(resolverClass, conf);
+      
+      // 4. Group files using custom grouper
+      Collection<Collection<Pair<String, Long>>> groupedFiles = 
+        customFileGrouper.getGroupedInputFiles(hfiles);
+      
+      // 5. Create InputSplits from grouped files
+      List<InputSplit> splits = new ArrayList<>();
+      
+      for (Collection<Pair<String, Long>> group : groupedFiles) {
+        // Get location hints for this group
+        Set<String> locations = fileLocationResolver.getLocationsForInputFiles(group);
+        String[] locationArray = locations.toArray(new String[0]);
+        
+        // Create balanced splits from this group
+        List<InputSplit> groupSplits = createBalancedSplits(group, locationArray);
+        splits.addAll(groupSplits);
+      }
+
+      LOG.info("Created {} rack-aware InputSplits from {} HFiles in {} groups", 
+        splits.size(), hfiles.size(), groupedFiles.size());
+      
+      return splits;
+    }
+    
+    private List<InputSplit> createBalancedSplits(Collection<Pair<String, Long>> hfiles, 
+        String[] locations) throws IOException {
+      List<InputSplit> splits = new ArrayList<>();
+      
+      // For now, create one split per HFile (can be enhanced for better balancing)
+      for (Pair<String, Long> hfile : hfiles) {
+        Path path = new Path(hfile.getFirst());
+        long length = hfile.getSecond();
+        if (length <= 0){
+          LOG.warn("Skipping empty or invalid Hfile: {} with length: {}", path, length);
+          continue;
+        }
+
+        splits.add(new FileSplit(path, 0, length, locations));
+      }
+      
+      return splits;
+    }
+  }
```

Comment:
How does HFileInputFormat group files into splits? Is it just one file per split?